docs/guide/rules-concept.md

```diff
@@ -0,0 +1,181 @@
+    }
+}
+```
+
+Note that third argument in `validate()` is an instance of `ValidationContext` so you can use it if you need
+whole data set context. For example, implementation might be the following if you need to validate "company"
+property only if "hasCompany" is true:
+
+```php
+namespace MyVendor\Rules;
+
+use Yiisoft\Validator\DataSetInterface;
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\Rule;
+use Yiisoft\Validator\ValidationContext;
+
+final class CompanyNameHandler implements Rule\RuleHandlerInterface
+{
+    use FormatMessageTrait;
+    
+    private FormatterInterface $formatter;
+
+    public function validate(mixed $value, object $rule, ?ValidationContext $context = null): Result
+    {
+        if (!$rule instanceof CompanyName) {
+            throw new UnexpectedRuleException(CompanyName::class, $rule);
+        }
+        
+        $result = new Result();
+        $dataSet = $context->getDataSet();
+        $hasCompany = $dataSet->getAttributeValue('hasCompany') === true;
+
+        if ($hasCompany && $this->isCompanyNameValid($value) === false) {
+            $result->addError('Company name is not valid.');
+        }
+
+        return $result;
+    }
+
+    private function isCompanyNameValid(string $value): bool
+    {
+        // check company name    
+    }
+}
+```
+
+> Note: Do not call handler's `validate()` method directly. It must be used via Validator only.
+
+##### Resolving rule handler dependencies
+
+Basically, you can use `SimpleRuleHandlerResolver` to resolve rule handler.
+In case you need extra dependencies, this can be done by `ContainerRuleHandlerResolver`.
+
+That would work with the following implementation:
+
+```php
+final class NoLessThanExistingBidRuleHandler implements RuleHandlerInterface
+{
+    use FormatMessageTrait;
+    
+    private FormatterInterface $formatter;
+
+    public function __construct(    
+        private ConnectionInterface $connection,        
+        ?FormatterInterface $formatter = null
+    ) {
+        $this->formatter = $formatter ?? new Formatter();
+    }
+    }
+    
+    public function validate(mixed $value, object $rule, ?ValidationContext $context): Result
+    {
+        $result = new Result();
+        
+        $currentMax = $connection->query('SELECT MAX(price) FROM bid')->scalar();
+        if ($value <= $currentMax) {
+            $result->addError($this->formatter->format('There is a higher bid of {bid}.', ['bid' => $currentMax]));
+        }
+
+        return $result;
+    }
+}
+
+$ruleHandlerContainer = new ContainerRuleHandlerResolver(new MyContainer());
+$ruleHandler = $ruleHandlerContainer->resolve(NoLessThanExistingBidRuleHandler::class);
+```
+
+`MyContainer` is a container for resolving dependencies and  must be an instance of
+`Psr\Container\ContainerInterface`. [Yii Dependency Injection](https://github.com/yiisoft/di) implementation also can
+be used.
+
+###### Using [Yii config](https://github.com/yiisoft/config)
+
+```php
```

Comment:
Need intro sentence.

Reply:
Will be fixed after finishing "Rules handlers resolvers" section.

Reply:
Suggest to add this section in a separate PR.